source/common/quic/envoy_quic_dispatcher.cc

```diff
@@ -126,6 +126,23 @@ std::unique_ptr<quic::QuicSession> EnvoyQuicDispatcher::CreateQuicSession(
   return quic_session;
 }
 
+bool EnvoyQuicDispatcher::OnFailedToDispatchPacket(
+    const quic::ReceivedPacketInfo& received_packet_info) {
+  if (hot_restart_forward_packet_ == nullptr) {
+    // If we're not hot restarting, do the default behavior.
+    return quic::QuicDispatcher::OnFailedToDispatchPacket(received_packet_info);
+  }
+  // We are hot restarting, forward the packet to the new instance.
+  Network::UdpRecvData packet;
+  packet.addresses_.local_ = quicAddressToEnvoyAddressInstance(received_packet_info.self_address);
+  packet.addresses_.peer_ = quicAddressToEnvoyAddressInstance(received_packet_info.peer_address);
+  packet.buffer_ = std::make_unique<Buffer::OwnedImpl>(received_packet_info.packet.AsStringPiece());
+  packet.receive_time_ = MonotonicTime(
+      std::chrono::microseconds{received_packet_info.packet.receipt_time().ToDebuggingValue()});
```

Comment:
Instead of converting a Network::UdpRecvData object in ActiveQuicListener::onDataWorker() to a quic::QuicReceivedPacket and then to quic::ReceivedPacketInfo and then convert it back to UdpRecvData, can we move packet forwarding logic to the listener? I.e. we can keep some states in the EnvoyQuicDispatcher about whether the Envoy is during hot restart and whether the current packet should forwarded. Then at end of onDataWorker() check the states and move the UdpRecvData if forwarding is needed.

Reply:
That is an option, but doing it that way adds extra checks in the common path (for every normal successfully delivered packet, not during hot restart). I think a small cost for packets that need to be forwarded is better than a tiny cost in the common path, yes/no?

Reply:
(Especially bearing in mind that ideally this whole thing is not in the flow at all if an eBPF implementation takes over, at which point the comparison would be a tiny cost in the common path if we check state in onDataWorker, versus zero cost of any kind.)

Reply:
I won't think checking a bool followed by no-op is a cost at all with eBPF. But forwarding packets in hot restarting has substantial cost, and if we copy every packets to be forwarded, that's even more cost.

Reply:
The forwarding is already much more expensive than one copy though, and, again, ideally in the performance-sensitive situation this will be completely moot with the eBPF version.
IMO a tiny cost (it's not zero, and it's a bit more than checking a bool, it's clearing and checking a bool every time in the common path, though at least the branch behavior will be mostly predictable) and additional complication in the common path, is a much worse compromise than almost any amount of complexity or slowdown in the occasional, rare, *and eventually unused* forwarding-during-hot-restart path. It's like a reference implementation of a graphics driver, it's important that it does the right thing, with performance as a distant secondary concern.

That said, if you feel really strongly about this, I can change it to how you want it. Having something that works is more important than having something that works the way I would prefer.

Reply:
If draining period would be long, the Envoy will be forwarding most of the incoming packets at the end. Reducing copy is definitely worth doing. It's up to you to decide if your system want this or not, since you know better about the traffic pattern.